clinics/merge/cluster.go

```diff
@@ -0,0 +1,235 @@
+package merge
+
+import (
+	"context"
+	"errors"
+	"github.com/dominikbraun/graph"
+	"github.com/eapache/queue"
+	"github.com/tidepool-org/clinic/patients"
+	"slices"
+)
+
+const (
+	PatientAttributeDOB      = "dob"
+	PatientAttributeFullName = "fullName"
+	PatientAttributeMRN      = "mrn"
+	PatientAttributeUserId   = "userId"
+
+	conflictAttributeKey = "conflict"
+)
+
+var (
+	attributeGetters = map[string]func(patient patients.Patient) string{
+		PatientAttributeDOB:      getDOB,
+		PatientAttributeFullName: getFullName,
+		PatientAttributeMRN:      getMRN,
+		PatientAttributeUserId:   getUserId,
+	}
+)
+
+type PatientClusters []PatientCluster
+
+func (p PatientClusters) PreventsMerge() bool {
+	return false
+}
+
+type PatientCluster struct {
+	Patients []PatientConflicts
+}
+
+type PatientConflicts struct {
+	// Source patient
+	Patient patients.Patient
+
+	// Conflicts is a map from conflict category to target patient user id
+	Conflicts map[string][]string
+}
+
+type PatientClusterReporter struct {
+	graph             graph.Graph[string, patients.Patient]
+	patients          []patients.Patient
+	targetByAttribute map[string]map[string][]*patients.Patient
+}
+
+func NewPatientClusterReporter(pts []patients.Patient) *PatientClusterReporter {
+	reporter := &PatientClusterReporter{
+		graph:             graph.New(getUserId),
+		patients:          pts,
+		targetByAttribute: buildAttributeMap(pts),
+	}
+
+	return reporter
+}
+
+func (p *PatientClusterReporter) Plan(ctx context.Context) (PatientClusters, error) {
+	return p.GetPatientClusters()
+}
+
+func (p *PatientClusterReporter) GetPatientClusters() (PatientClusters, error) {
+	for _, patient := range p.patients {
+		if err := p.graph.AddVertex(patient); err != nil {
+			return nil, err
+		}
+	}
+	for _, patient := range p.patients {
+		if err := p.addDuplicateEdges(patient); err != nil {
+			return nil, err
+		}
+	}
+
+	visited := map[string]struct{}{}
+	clusters := make([]PatientCluster, 0)
+	adjecencyMap, err := p.graph.AdjacencyMap()
+	if err != nil {
+		return nil, err
+	}
+
+	// BFS traversal to get each patient cluster
+	for userId := range adjecencyMap {
+		cluster := PatientCluster{}
+		q := queue.New()
+		q.Add(userId)
+		for q.Length() != 0 {
+			id := q.Remove().(string)
+			if _, ok := visited[id]; ok {
+				continue
+			}
+
+			patient, err := p.graph.Vertex(id)
+			if err != nil {
+				return nil, err
+			}
+
+			conflicts := map[string][]string{}
+			for duplicate, edge := range adjecencyMap[id] {
+				q.Add(duplicate)
+				conflict := edge.Properties.Attributes[conflictAttributeKey]
+				conflicts[conflict] = append(conflicts[conflict], duplicate)
```

Comment:
For my understanding, if user `"A`" had a `"MRN Only Match`" conflict with user `"B`", and `"B`" has a `"Name Only Match`" conflict with user `"C`", `"A`", `"B`", and `"C`" would be part of the same cluster right?

Reply:
Correct